File: src/lib/calendar/CalendarGrid.ts
/**
 * Grade do calendário (semanas × dias) — exatamente uma matriz 6×7.
 *
 * Conceitos AP CS A:
 *  - Unit 8 (2D Arrays): a grade é um int[][] em Java, ou
 *    DayCell[][] aqui. Acessamos via grid[row][col]; varremos com
 *    nested for. Isso é o coração da Unit 8.
 *  - Unit 4 (Iteration): nested loops para popular e renderizar.
 *  - Unit 1/2 (Primitive types, math): cálculo da primeira coluna usa
 *    Date.getDay() e aritmética de inteiros.
 *
 * Layout: linhas = semanas (0..5), colunas = dias da semana
 * (0=domingo, ..., 6=sábado). Padrão americano/brasileiro idêntico.
 */

export type DayCell = {
  // Os três campos abaixo identificam a data calendárica do slot.
  year: number
  month: number // 0-indexed (igual Date)
  day: number
  // Se este slot pertence ao mês corrente ou ao "spillover" (mês anterior
  // / posterior aparecendo nas bordas da grade).
  inMonth: boolean
}

export class CalendarGrid {
  // Unit 8: ARRAY 2D — semanas × dias da semana.
  private readonly cells: DayCell[][]
  private readonly year: number
  private readonly month: number // 0-indexed

  public constructor(year: number, month: number) {
    this.year = year
    this.month = month
    this.cells = CalendarGrid.build(year, month)
  }

  public getYear(): number {
    return this.year
  }
  public getMonth(): number {
    return this.month
  }

  // O AP gosta MUITO de "row-major traversal".
  public getCell(row: number, col: number): DayCell {
    return this.cells[row][col]
  }

  // Defensive copy do array 2D.
  public getCells(): DayCell[][] {
    return this.cells.map((row) => row.slice())
  }

  public getRowCount(): number {
    return this.cells.length
  }
  public getColCount(): number {
    return this.cells[0]?.length ?? 0
  }

  // Unit 4 + 8: nested loop para encontrar a célula de uma data.
  // Retorna [row, col] ou null. Demonstra "row-major search".
  public findCell(
    year: number,
    month: number,
    day: number,
  ): [number, number] | null {
    for (let r = 0; r < this.cells.length; r++) {
      for (let c = 0; c < this.cells[r].length; c++) {
        const cell = this.cells[r][c]
        if (
          cell.year === year &&
          cell.month === month &&
          cell.day === day
        ) {
          return [r, c]
        }
      }
    }
    return null
  }

  // Constrói a matriz 6×7 que sempre cobre o mês inteiro.
  // Lógica: começa no domingo da semana que contém o dia 1, e enche
  // 42 slots seguidos. Os slots fora do mês corrente recebem inMonth=false.
  private static build(year: number, month: number): DayCell[][] {
    const first = new Date(year, month, 1)
    const startWeekday = first.getDay() // 0..6
    // Volta `startWeekday` dias para chegar no domingo dessa semana.
    const start = new Date(year, month, 1 - startWeekday)

    const grid: DayCell[][] = []
    const cursor = new Date(start)

    for (let r = 0; r < 6; r++) {
      const row: DayCell[] = []
      for (let c = 0; c < 7; c++) {
        row.push({
          year: cursor.getFullYear(),
          month: cursor.getMonth(),
          day: cursor.getDate(),
          inMonth: cursor.getMonth() === month,
        })
        cursor.setDate(cursor.getDate() + 1)
      }
      grid.push(row)
    }
    return grid
  }

  // Helper estático: nome do mês em PT-BR.
  public static monthName(month: number): string {
    const names = [
      "Janeiro",
      "Fevereiro",
      "Março",
      "Abril",
      "Maio",
      "Junho",
      "Julho",
      "Agosto",
      "Setembro",
      "Outubro",
      "Novembro",
      "Dezembro",
    ]
    return names[month] ?? ""
  }

  // Cabeçalhos das colunas (segunda, terça, ...).
  // Mantemos domingo na primeira coluna porque é o que combina com
  // Date.getDay() — se quiser semana começando segunda, basta rotacionar.
  public static weekdayHeaders(): readonly string[] {
    return ["Dom", "Seg", "Ter", "Qua", "Qui", "Sex", "Sáb"]
  }
}

File: java/CalendarGrid.java
/*
 * CalendarGrid — grade do calendario (6 semanas x 7 dias).
 *
 * Unit 8 — ARRAY 2D: DayCell[][]. Acesso via grid[row][col].
 * Unit 4 — Nested loops pra varrer / buscar.
 */
public class CalendarGrid {

    // Unit 8: ARRAY 2D. 6 linhas (semanas) x 7 colunas (dias da semana).
    private final DayCell[][] cells;
    private final int year;
    private final int month;

    public CalendarGrid(int year, int month) {
        this.year = year;
        this.month = month;
        this.cells = build(year, month);
    }

    public int getYear()         { return year; }
    public int getMonth()        { return month; }
    public int getRowCount()     { return cells.length; }
    public int getColCount()     { return (cells.length == 0) ? 0 : cells[0].length; }
    public DayCell getCell(int row, int col) { return cells[row][col]; }

    public DayCell[][] getCells() {
        // Defensive copy de array 2D — copia linha por linha.
        DayCell[][] out = new DayCell[cells.length][];
        for (int r = 0; r < cells.length; r++) {
            out[r] = new DayCell[cells[r].length];
            for (int c = 0; c < cells[r].length; c++) {
                out[r][c] = cells[r][c];
            }
        }
        return out;
    }

    // Unit 4 + 8: nested loop pra varrer a matriz inteira buscando
    // uma data especifica. Demonstra "row-major search".
    public int[] findCell(int y, int m, int d) {
        for (int r = 0; r < cells.length; r++) {
            for (int c = 0; c < cells[r].length; c++) {
                DayCell cell = cells[r][c];
                if (cell.getYear() == y && cell.getMonth() == m
                        && cell.getDay() == d) {
                    return new int[] { r, c };
                }
            }
        }
        return null; // nao encontrado
    }

    // Unit 5 + 8: utility privado pra montar a matriz.
    // 42 slots (6*7). Comeca no domingo da semana que contem o dia 1.
    private static DayCell[][] build(int year, int month) {
        // weekday do dia 1: 0=domingo .. 6=sabado.
        int startWeekday = zellersWeekday(year, month, 1);

        // Volta startWeekday dias a partir do dia 1.
        int cursorYear = year;
        int cursorMonth = month;
        int cursorDay = 1 - startWeekday;
        // Normaliza se ficou negativo (mes anterior).
        while (cursorDay <= 0) {
            cursorMonth--;
            if (cursorMonth < 1) { cursorMonth = 12; cursorYear--; }
            cursorDay = cursorDay + daysInMonth(cursorYear, cursorMonth);
        }

        DayCell[][] grid = new DayCell[6][7];
        for (int r = 0; r < 6; r++) {
            for (int c = 0; c < 7; c++) {
                boolean inMonth = (cursorYear == year && cursorMonth == month);
                grid[r][c] = new DayCell(cursorYear, cursorMonth, cursorDay, inMonth);
                // proximo dia
                cursorDay++;
                if (cursorDay > daysInMonth(cursorYear, cursorMonth)) {
                    cursorDay = 1;
                    cursorMonth++;
                    if (cursorMonth > 12) { cursorMonth = 1; cursorYear++; }
                }
            }
        }
        return grid;
    }

    // Zeller's congruence (simplificada): retorna 0=Dom..6=Sab.
    // Vale pro calendario gregoriano.
    public static int zellersWeekday(int y, int m, int d) {
        int yy = y;
        int mm = m;
        if (mm < 3) { mm = mm + 12; yy = yy - 1; }
        int K = yy % 100;
        int J = yy / 100;
        int h = (d + (13 * (mm + 1)) / 5 + K + K/4 + J/4 + 5*J) % 7;
        // h: 0=sabado, 1=domingo... Convertemos pra 0=dom..6=sab.
        int weekday = (h + 6) % 7;
        return weekday;
    }

    public static int daysInMonth(int year, int month) {
        if (month == 4 || month == 6 || month == 9 || month == 11) return 30;
        if (month == 2) {
            // Bissexto: divisivel por 4, exceto seculos nao-divisiveis
            // por 400.
            boolean leap = (year % 4 == 0 && year % 100 != 0)
                         || (year % 400 == 0);
            return leap ? 29 : 28;
        }
        return 31;
    }

    public static String monthName(int month) {
        String[] names = {
            "Janeiro", "Fevereiro", "Marco", "Abril",
            "Maio", "Junho", "Julho", "Agosto",
            "Setembro", "Outubro", "Novembro", "Dezembro"
        };
        if (month < 1 || month > 12) return "";
        return names[month - 1];
    }

    public static String[] weekdayHeaders() {
        return new String[] {
            "Dom", "Seg", "Ter", "Qua", "Qui", "Sex", "Sab"
        };
    }
}
